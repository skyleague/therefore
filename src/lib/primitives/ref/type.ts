/* eslint-disable @typescript-eslint/no-unsafe-declaration-merging */
import type { NodeTrait } from '../../cst/mixin.js'
import { Node } from '../../cst/node.js'
import type { Intrinsic } from '../../cst/types.js'
import type { SchemaOptions } from '../base.js'

import { type ConstExpr, evaluate } from '@skyleague/axioms'

export type RefOptions = object

export class RefType<Reference extends Node = Node> extends Node {
    public override _type = 'ref' as const
    public override _children: [Node]

    public _options: RefOptions = {}
    public declare infer: Reference['infer']
    public declare input: Reference['input']
    public declare intrinsic: Intrinsic<Reference>

    public constructor(reference: ConstExpr<Reference>, options: SchemaOptions<RefOptions, Reference['infer']>) {
        super(options)
        this._options = options

        // we are lying to typescript here, but it's fine
        // there expression will be evaluated in the prepass, and errors hard if it's not
        this._children = [reference as Node]
    }

    public override _hooks = {
        onLoad: [
            (self: Node): void => {
                self._children = self._children?.map(evaluate) as [Node]
            },
        ],
        onGenerate: [
            ({ _children: children }: Node): void => {
                const [reference] = children ?? []
                if (reference !== undefined && reference._sourcePath === undefined) {
                    // this reference was not exported on its own
                    // so we just call it local if has no name
                    reference._name ??= 'local'
                }
            },
        ],
    }

    public get item(): Reference {
        return this._children[0] as Reference
    }

    public static _from<Reference extends Node>(
        ref: ConstExpr<Reference>,
        options: SchemaOptions<RefOptions>,
    ): RefType<Intrinsic<Reference>> {
        type IntrinsicType = Intrinsic<Reference>
        return new RefType<IntrinsicType>(ref as ConstExpr<IntrinsicType>, options)
    }
}

export interface RefType extends Node, NodeTrait {}
