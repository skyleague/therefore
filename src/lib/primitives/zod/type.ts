import type { ArrayType } from '../array/array.js'
import type { BooleanType } from '../boolean/boolean.js'
import type { ConstType } from '../const/const.js'
import type { IntersectionType } from '../intersection/intersection.js'
import type { NullableType } from '../nullable/nullable.js'
import type { NumberType } from '../number/number.js'
import type { ObjectType } from '../object/object.js'
import type { OptionalType } from '../optional/optional.js'
import type { RecordType } from '../record/record.js'
import type { StringType } from '../string/string.js'
import type { TupleType } from '../tuple/tuple.js'
import type { DiscriminatedUnionType, UnionType } from '../union/union.js'
import type { UnknownType } from '../unknown/unknown.js'

export type ZodSchema = { _def: { typeName: unknown }; _output: unknown }

export type ZodSchemaAsNode<T extends ZodSchema> = T['_def']['typeName'] extends 'ZodString'
    ? StringType
    : T['_def']['typeName'] extends 'ZodNumber'
      ? NumberType
      : T['_def']['typeName'] extends 'ZodBoolean'
        ? BooleanType
        : T['_def']['typeName'] extends 'ZodUndefined'
          ? ConstType<undefined>
          : T['_def']['typeName'] extends 'ZodNull'
            ? ConstType<null>
            : T['_def']['typeName'] extends 'ZodAny'
              ? UnknownType
              : T['_def']['typeName'] extends 'ZodUnknown'
                ? UnknownType
                : T extends { _def: { typeName: 'ZodArray' }; element: infer Element }
                  ? Element extends ZodSchema
                      ? ArrayType<ZodSchemaAsNode<Element>>
                      : never
                  : T extends { _def: { typeName: 'ZodObject' }; shape: infer Shape }
                    ? ObjectType<{
                          [K in keyof Shape]: Shape[K] extends ZodSchema ? ZodSchemaAsNode<Shape[K]> : never
                      }>
                    : T['_def']['typeName'] extends 'ZodUnion'
                      ? UnionType
                      : T['_def']['typeName'] extends 'ZodDiscriminatedUnion'
                        ? DiscriminatedUnionType
                        : T['_def']['typeName'] extends 'ZodIntersection'
                          ? IntersectionType
                          : T['_def']['typeName'] extends 'ZodTuple'
                            ? TupleType
                            : T extends { _def: { typeName: 'ZodRecord' }; element: infer Element }
                              ? RecordType<Element extends ZodSchema ? ZodSchemaAsNode<Element> : never>
                              : T extends { _def: { typeName: 'ZodLiteral' }; value: infer Value }
                                ? ConstType<Value>
                                : T extends { _def: { typeName: 'ZodOptional'; innerType: infer Schema } }
                                  ? Schema extends ZodSchema
                                      ? OptionalType<ZodSchemaAsNode<Schema>>
                                      : never
                                  : T extends { _def: { typeName: 'ZodNullable'; innerType: infer Schema } }
                                    ? Schema extends ZodSchema
                                        ? NullableType<ZodSchemaAsNode<Schema>>
                                        : never
                                    : T['_def']['typeName'] extends 'ZodDefault'
                                      ? BooleanType //
                                      : T extends {
                                              _def: { typeName: 'ZodEffects' }
                                              innerType: (...args: unknown[]) => infer Schema
                                          }
                                        ? Schema extends ZodSchema
                                            ? ZodSchemaAsNode<Schema>
                                            : never
                                        : never
