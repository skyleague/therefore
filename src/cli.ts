import type { CommandModule } from 'yargs'
import yargs from 'yargs'
import { hideBin } from 'yargs/helpers'
import packageJSON from '../package.json' with { type: 'json' }
import * as commands from './commands/index.js'

const { bin } = packageJSON

/**
 * @internal
 */
export async function run(): Promise<void> {
    let cli = yargs(hideBin(process.argv)).scriptName(Object.keys(bin)[0] ?? 'cli')
    for (const command of Object.values(commands)) {
        cli = cli.command(command.default as unknown as CommandModule)
    }
    await cli.demandCommand().strict().help().argv
}
