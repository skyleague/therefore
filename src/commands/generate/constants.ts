import { name } from '../../../package.json'

export const generatedBy = `Generated by ${name}`
export const generatedAjv = `Generated by Ajv https://ajv.js.org/guide/managing-schemas.html#standalone-validation-code`
