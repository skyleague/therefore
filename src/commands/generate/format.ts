import { exec } from 'node:child_process'
import { promisify } from 'node:util'
import { memoize } from '@skyleague/axioms'
import type { ThereforeOutputType } from '../../lib/output/types.js'

// biome-ignore lint/suspicious/noExplicitAny: it's a third-party library
export type Prettier = any
export async function maybeLoadPrettier() {
    try {
        // @ts-ignore
        return await import('prettier')
    } catch {
        return undefined
    }
}

export async function formatContent({
    prettier,
    input,
    file,
    type,
}: {
    prettier: Prettier
    input: string
    file: string
    type: ThereforeOutputType
}): Promise<string> {
    if (prettier !== undefined) {
        const options = await prettier.resolveConfig(file)
        return prettier.format(input, {
            ...(options ?? {}),
            parser: type,
        })
    }
    return input
}

const execAsync = promisify(exec)

const getBiomeBinaryPath = memoize(async () => {
    try {
        const { stdout } = await execAsync('npx --no-install biome --version')
        if (stdout.trim()) {
            return 'npx'
        }
    } catch (error) {
        console.error('Error finding Biome:', error)
    }
    return null
})

export async function formatBiomeFiles(files: string[]) {
    const biomePath = await getBiomeBinaryPath()
    if (!biomePath) {
        console.warn('Biome not found. Skipping formatting.')
        return
    }

    const baseCommand = `${biomePath} biome check --write `
    let fileChunks: string[][]
    if (process.platform === 'win32') {
        // Windows command line length limit
        const MAX_COMMAND_LENGTH = 8180
        const remainingLength = MAX_COMMAND_LENGTH - baseCommand.length

        fileChunks = []
        let currentChunk: string[] = []
        let currentLength = 0

        for (const file of files) {
            // +1 for the space between files
            // +2 for the quotes
            const fileLength = file.length + 1 + 2

            if (currentLength + fileLength > remainingLength) {
                fileChunks.push(currentChunk)
                currentChunk = []
                currentLength = 0
            }

            currentChunk.push(`"${file}"`)
            currentLength += fileLength
        }

        if (currentChunk.length > 0) {
            fileChunks.push(currentChunk)
        }
    } else {
        // On non-Windows platforms, process all files at once
        fileChunks = [files]
    }

    try {
        const promises = fileChunks.map(async (fileChunk) => {
            const command = `${baseCommand} ${fileChunk.join(' ')}`
            const { stdout, stderr } = await execAsync(command)

            if (stderr) {
                console.warn('Biome formatting warning:', stderr)
            }
            return stdout
        })

        const results = await Promise.all(promises)
        console.log('Biome formatting completed:', results.join('\n'))
    } catch (error) {
        console.error('Error running Biome format:', error)
    }
}
