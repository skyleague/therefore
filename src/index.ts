export type { OpenapiV3 } from './types/openapi.type.js'
export * from './lib/visitor/index.js'
export type { InferSchemaType, Schema, Parser } from './lib/types.js'
export * from './lib/cst/index.js'
export * from './lib/primitives/index.js'
