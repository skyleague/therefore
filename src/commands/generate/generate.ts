import { generatedAjv, generatedBy } from './constants.js'
import { expandGlobs } from './glob.js'
import { formatFile, maybeLoadPrettier } from './prettier.js'
import { resolveTypescriptSchema } from './resolver.js'
import type { FileDefinition, OutputFile, ReferenceData, ThereforeOutputType } from './types.js'

import { getExtension, replaceExtension } from '../../common/template/path.js'
import { renderTemplate } from '../../common/template/template.js'
import type { ThereforeNode } from '../../lib/cst/cst.js'
import type { ThereforeCst } from '../../lib/primitives/types.js'
import { isThereforeExport } from '../../lib/primitives/types.js'
import { toJsonSchema } from '../../lib/visitor/jsonschema/jsonschema.js'
import { prepass } from '../../lib/visitor/prepass/prepass.js'
import { toTypescriptDefinition } from '../../lib/visitor/typescript/typescript.js'

import { entriesOf, enumerate, groupBy, isDefined, range, second } from '@skyleague/axioms'
import decamelize from 'decamelize'

import fs from 'node:fs'
import path from 'node:path'
import { fileURLToPath } from 'node:url'

const __dirname = path.dirname(fileURLToPath(import.meta.url))
async function requireModule(module: string): Promise<Record<string, ThereforeNode | unknown>> {
    const relative = path.relative(__dirname, module).replace(/\\/g, '/').replace(/\.ts$/, '.js')

    const mod = (await import(relative.startsWith('.') ? relative : `./${relative}`)) as Record<string, unknown>
    return (mod.default ?? mod) as Record<string, ThereforeNode | unknown>
}

export async function loadSymbol({
    srcPath,
    basePath,
    symbol,
    sourceSymbol,
    compile,
    definitions,
    entry,
    fileSuffix,
    filePath,
    outputFileRename,
    appendSourceSymbolName: appendSourceName = false,
    rootPath = srcPath,
}: {
    srcPath: string
    basePath: string
    symbol: ThereforeCst
    sourceSymbol: string
    compile: boolean
    definitions: Record<string, FileDefinition | undefined>
    entry: string
    fileSuffix?: string | undefined
    filePath?: string | undefined
    outputFileRename: (path: string) => string
    appendSourceSymbolName?: boolean
    rootPath?: string
}) {
    const simplified = prepass(symbol)
    let targetDir = '.'
    if (simplified.type === 'custom') {
        if (simplified.value.fileSuffix !== undefined) {
            fileSuffix = simplified.value.fileSuffix
        }
        if (simplified.value.filePath !== undefined) {
            const dir = path.dirname(simplified.value.filePath)
            const basename = path.basename(simplified.value.filePath)
            filePath = basename
            targetDir = dir
        }
    }

    const { base: baseName, dir: baseDir } = path.parse(srcPath)
    const ext = getExtension(srcPath) ?? ''
    const targetBaseDir = path.relative(basePath, path.join(baseDir, targetDir))
    const targetSrcPath = `${targetBaseDir}/${(filePath ?? baseName).replace(ext, '')}${ext}`
    const targetPath = fileSuffix !== undefined ? replaceExtension(targetSrcPath, fileSuffix) : outputFileRename(targetSrcPath)

    definitions[targetPath] ??= {
        srcPath: path.relative(basePath, rootPath),
        targetPath,
        attachedFiles: [],
        symbols: [],
        dependencies: {},
    }

    const file = definitions[targetPath]!
    // check if symbol is not seen yet
    if (file.symbols.find((s) => s.uuid === simplified.uuid)) {
        return
    }
    const symbolName = simplified.name ?? sourceSymbol

    sourceSymbol = appendSourceName && simplified.name !== undefined ? `${sourceSymbol}.${simplified.name}` : symbolName
    const schemaName = decamelize(symbolName, { separator: '-' })
    const schemaFile = `./schemas/${schemaName}.schema.json`
    const compiledFile = `./schemas/${schemaName}.schema.js`

    const nodeIsCompiled = simplified.description.validator?.compile ?? compile

    const jsonschema = await toJsonSchema(simplified, nodeIsCompiled)
    const { definition, subtrees } = toTypescriptDefinition({ sourceSymbol, symbolName, schema: simplified })

    file.symbols.push(
        ...Object.values(definition.locals ?? {})
            .filter(isDefined)
            .map((local) => ({
                uuid: local.uuid,
                symbolName: local.symbolName,
                definition: local,
                typeOnly: true,
            }))
    )

    file.symbols.push({
        uuid: simplified.uuid,
        symbolName,
        definition: definition,
        schemaFile,
        compiledFile: nodeIsCompiled ? compiledFile : undefined,
        typeOnly: simplified.description.validator?.enabled !== true,
    })

    if (definition.isExported && simplified.description.validator?.enabled) {
        const cleanSchemasFolder = () => {
            const targetFolder = path.dirname(path.join(targetBaseDir, compiledFile))
            if (fs.existsSync(targetFolder)) {
                console.warn(`Cleaning ${targetFolder}`)
                fs.rmSync(targetFolder, { force: true, recursive: true })
            }
        }
        if (jsonschema.compiled) {
            file.attachedFiles.push({
                targetPath: path.join(targetBaseDir, compiledFile),
                content: `/**\n * ${generatedAjv} \n * eslint-disable\n */\n${jsonschema.code}`,
                prettify: false,
                type: 'validator',
                clean: cleanSchemasFolder,
            })
        } else {
            file.attachedFiles.push({
                targetPath: path.join(targetBaseDir, schemaFile),
                content: JSON.stringify(jsonschema.schema, null, 2),
                prettify: true,
                type: 'jsonschema',
                clean: cleanSchemasFolder,
            })
        }

        console.debug(` - found ${definition.symbolName}`)
    }
    for (const { node: subSymbol, fileSuffix: subFileSuffix, filePath: subFilePath } of subtrees) {
        await loadSymbol({
            symbol: subSymbol,
            basePath,
            sourceSymbol,
            compile,
            definitions,
            srcPath: path.join(basePath, targetPath),
            rootPath: srcPath,
            fileSuffix: subFileSuffix,
            filePath: subFilePath,
            entry,
            outputFileRename,
            appendSourceSymbolName: true,
        })
    }
}

export async function scanModule({
    entry,
    srcPath,
    basePath,
    compile,
    definitions,
    require = requireModule,
    outputFileRename,
}: {
    entry: string
    srcPath: string
    basePath: string
    compile: boolean
    definitions: Record<string, FileDefinition>
    require?: (module: string) => Promise<Record<string, ThereforeNode | unknown>>
    outputFileRename: (path: string) => string
}) {
    const module = await require(entry)

    for (const [sourceSymbol, symbolPromise] of Object.entries(module)) {
        const symbol = await symbolPromise

        if (!isThereforeExport(symbol)) {
            continue
        }

        await loadSymbol({ symbol, sourceSymbol, compile, definitions, srcPath, entry, basePath, outputFileRename })
    }
}

export async function scanFiles({
    files,
    basePath,
    compile,
    outputFileRename,
}: {
    files: string[]
    basePath: string
    compile: boolean
    outputFileRename: (path: string) => string
}): Promise<Record<string, FileDefinition>> {
    const definitions: Record<string, FileDefinition> = {}
    for (const entry of files) {
        const srcPath = path.resolve(basePath, entry)
        let isGenerated = false
        const strippedPath = srcPath.substr(0, srcPath.lastIndexOf('.'))
        for (const ext of ['.ts', '.js']) {
            const p = `${strippedPath}${ext}`
            isGenerated ||= fs.existsSync(p) && fs.readFileSync(p).includes(generatedBy)
        }
        if (isGenerated) {
            console.debug(`scanning ${entry}`)
            console.debug(` * skipping generated schema`)
            continue
        }

        console.log(`scanning ${entry}`)
        try {
            await scanModule({ entry, basePath, srcPath, compile, definitions, outputFileRename })
        } catch (e: unknown) {
            const error = e as Error
            console.debug(error.message, error.stack)
            throw error
        }
    }
    return definitions
}

export async function compileOutputFiles(
    entries: string[],
    { compile = false, cwd, outputFileRename }: { cwd: string; compile: boolean; outputFileRename: (path: string) => string }
): Promise<
    {
        targetPath: string
        template: string
        data: Record<string, ReferenceData>
        type: ThereforeOutputType
        prettify: boolean
        clean?: () => void
    }[]
> {
    const outputFiles: OutputFile[] = []
    const definitions = await scanFiles({ files: entries, basePath: cwd, compile, outputFileRename })

    const localReferences = {}

    for (const targetPath of Object.keys(definitions)) {
        resolveTypescriptSchema({ definitions, targetPath, cwd, outputFiles, localReferences })
    }

    return compile ? outputFiles : outputFiles.filter((f) => f.type !== 'validator')
}

export async function generate({
    globs,
    ignore,
    extension,
    compile,
    outputFileRename,
    clean,
}: {
    globs: string[]
    ignore: string[]
    extension: string
    compile: boolean
    outputFileRename: (path: string) => string
    clean: boolean
}): Promise<void> {
    const cwd = process.cwd()
    const entries = await expandGlobs({ patterns: globs, ignore: [`!${extension}`, ...ignore], cwd, extension })

    const outputFiles = await compileOutputFiles(entries, {
        cwd,
        outputFileRename,
        compile,
    })

    if (clean) {
        for (const of of outputFiles) {
            if (of.clean !== undefined) {
                of.clean()
            }
        }
    }

    const prettier = maybeLoadPrettier()

    const data = outputFiles.reduce<Record<string, ReferenceData>>(
        (agg, x) => ({
            ...agg,
            ...x.data,
        }),
        {}
    )

    const duplicates = entriesOf(groupBy(entriesOf(data), ([, values]) => `${values.srcPath}${values.symbolName}`)).filter(
        ([, records]) => records.length > 1
    )
    for (const duplicateGroup of duplicates.map((d) => second(d))) {
        for (const [i, [key]] of enumerate(duplicateGroup)) {
            // data[key].referenceName = `${data[key].referenceName}${i}`
            data[key]!.symbolName = `${data[key]!.symbolName}${i}`
            data[key]!.uniqueSymbolName = `${data[key]!.uniqueSymbolName}${i}`
        }
    }

    const variablesMap = Object.fromEntries(
        Object.entries(data)
            .map(([key, vals]) => Object.entries(vals).map(([varName, value]) => [`${key}:${varName}`, value] as const))
            .flat()
    )

    for (const { targetPath, template, type, prettify } of outputFiles) {
        fs.mkdirSync(path.dirname(targetPath), { recursive: true })
        // allow nested references up to 3 deep
        let contents = template
        for (const _ of range(3)) {
            contents = renderTemplate(contents, variablesMap)
        }
        fs.writeFileSync(targetPath, prettify ? await formatFile(prettier, contents, targetPath, type) : contents)
    }
}
