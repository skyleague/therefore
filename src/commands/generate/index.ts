import { generate } from './generate'

import { getExtension } from '../../common/template/path'

import type { Argv } from 'yargs'

export function builder(yargs: Argv) {
    return yargs
        .option('files', {
            alias: 'f',
            demandOption: true,
            describe: 'globs to scan for schemas',
            type: 'array',
        })
        .option('ignore-pattern', {
            alias: 'i',
            demandOption: true,
            default: ['**/*.d.ts', 'node_modules'],
            describe: 'globs to exclude',
            type: 'array',
        })
        .option('compile', {
            default: true,
            type: 'boolean',
        })
        .option('ext', {
            default: '.schema.ts',
            type: 'string',
        })
        .option('out-ext', {
            default: '.type.ts',
            type: 'string',
        })
        .option('clean', {
            default: false,
            type: 'boolean',
        })
}

export async function handler(argv: ReturnType<typeof builder>['argv']): Promise<void> {
    const { files = [], 'ignore-pattern': ignorePatterns, ext, 'out-ext': outExt, compile, clean } = await argv

    await generate({
        globs: files.map((f) => f.toString()),
        ignore: ignorePatterns.map((i) => i.toString()),
        extension: ext,
        outputFileRename: (path: string) => path.replace(getExtension(path)!, outExt),
        compile,
        clean,
    })
}

export default {
    command: ['generate', '*'],
    describe: 'generate therefore types',
    builder,
    handler,
}
