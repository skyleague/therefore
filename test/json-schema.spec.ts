import { SalesPerson } from '../examples/json-schema/example.type'
import { arbitrary } from '../src'
import { compileOutputFiles } from '../src/commands/generate/generate'

import { forAll } from '@skyleague/axioms'

test('intersection', async () => {
    forAll(await arbitrary(SalesPerson), (x) => SalesPerson.is(x))
})

test('output generation', async () => {
    expect(
        await compileOutputFiles(['examples/json-schema/example.schema.ts'], {
            outputFileRename: (file: string) => file.replace('.ts', '.type.ts'),
            cwd: process.cwd(),
            compile: true,
        })
    ).toMatchSnapshot()
})
