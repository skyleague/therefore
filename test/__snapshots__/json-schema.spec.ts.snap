// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`output generation 1`] = `
[
  {
    "data": {
      "00014-000": {
        "referenceName": "{{00014-000:symbolName}}",
        "srcPath": "examples/json-schema/example.schema.type.ts",
        "symbolName": "SelfReference",
        "uniqueSymbolName": "{{00014-000:symbolName}}",
      },
      "0005-000": {
        "referenceName": "{{0005-000:symbolName}}",
        "srcPath": "examples/json-schema/example.schema.type.ts",
        "symbolName": "Person",
        "uniqueSymbolName": "{{0005-000:symbolName}}",
      },
      "0009-000": {
        "referenceName": "{{0009-000:symbolName}}",
        "srcPath": "examples/json-schema/example.schema.type.ts",
        "symbolName": "SalesPerson",
        "uniqueSymbolName": "{{0009-000:symbolName}}",
      },
    },
    "prettify": true,
    "relativeSource": "./example.schema",
    "targetPath": "examples/json-schema/example.schema.type.ts",
    "template": "/**
* Generated by @skyleague/therefore@v1.0.0-local
* Do not manually touch this
*/
/* eslint-disable */
import AjvValidator from 'ajv'
import type { ValidateFunction } from 'ajv'


export interface {{0005-000:symbolName}} {
    /**
     * The person's first name.
     */
    firstName: string
    lastName: string
    age: number
}

export const {{0005-000:symbolName}} = {
    validate: require('./schemas/person.schema.js') as ValidateFunction<{{0005-000:symbolName}}>,
    get schema() { return {{0005-000:symbolName}}.validate.schema},
    is: (o: unknown): o is {{0005-000:symbolName}} => {{0005-000:symbolName}}.validate(o) === true,
    assert: (o: unknown) => {
        if (!{{0005-000:symbolName}}.validate(o)) {
            throw new AjvValidator.ValidationError({{0005-000:symbolName}}.validate.errors ?? [])
        }
    },
} as const

export type {{0009-000:symbolName}} = {
    sales: number
} & {
    /**
     * The person's first name.
     */
    firstName: string
    lastName: string
    age: number
}

export const {{0009-000:symbolName}} = {
    validate: require('./schemas/sales-person.schema.js') as ValidateFunction<{{0009-000:symbolName}}>,
    get schema() { return {{0009-000:symbolName}}.validate.schema},
    is: (o: unknown): o is {{0009-000:symbolName}} => {{0009-000:symbolName}}.validate(o) === true,
    assert: (o: unknown) => {
        if (!{{0009-000:symbolName}}.validate(o)) {
            throw new AjvValidator.ValidationError({{0009-000:symbolName}}.validate.errors ?? [])
        }
    },
} as const

export interface {{00014-000:symbolName}} {
    foo?: string
    bar?: {{00014-000:referenceName}}
}

",
    "type": "typescript",
  },
  {
    "clean": [Function],
    "data": {
      "00014-000": {
        "referenceName": "{{00014-000:symbolName}}",
        "srcPath": "examples/json-schema/example.schema.type.ts",
        "symbolName": "SelfReference",
        "uniqueSymbolName": "{{00014-000:symbolName}}",
      },
      "0005-000": {
        "referenceName": "{{0005-000:symbolName}}",
        "srcPath": "examples/json-schema/example.schema.type.ts",
        "symbolName": "Person",
        "uniqueSymbolName": "{{0005-000:symbolName}}",
      },
      "0009-000": {
        "referenceName": "{{0009-000:symbolName}}",
        "srcPath": "examples/json-schema/example.schema.type.ts",
        "symbolName": "SalesPerson",
        "uniqueSymbolName": "{{0009-000:symbolName}}",
      },
    },
    "prettify": false,
    "relativeSource": "./example.schema",
    "targetPath": "examples/json-schema/schemas/person.schema.js",
    "template": "/**
 * Generated by @skyleague/therefore
 * eslint-disable
 */
"use strict";module.exports = validate10;module.exports.default = validate10;const schema11 = {"$schema":"http://json-schema.org/draft-07/schema#","title":"{{0005-000:uniqueSymbolName}}","type":"object","properties":{"firstName":{"type":"string","description":"The person's first name."},"lastName":{"type":"string"},"age":{"type":"number"}},"required":["firstName","lastName","age"],"additionalProperties":false};function validate10(data, {instancePath="", parentData, parentDataProperty, rootData=data}={}){let vErrors = null;let errors = 0;if(errors === 0){if(data && typeof data == "object" && !Array.isArray(data)){let missing0;if((((data.firstName === undefined) && (missing0 = "firstName")) || ((data.lastName === undefined) && (missing0 = "lastName"))) || ((data.age === undefined) && (missing0 = "age"))){validate10.errors = [{instancePath,schemaPath:"#/required",keyword:"required",params:{missingProperty: missing0},message:"must have required property '"+missing0+"'"}];return false;}else {const _errs1 = errors;for(const key0 in data){if(!(((key0 === "firstName") || (key0 === "lastName")) || (key0 === "age"))){validate10.errors = [{instancePath,schemaPath:"#/additionalProperties",keyword:"additionalProperties",params:{additionalProperty: key0},message:"must NOT have additional properties"}];return false;break;}}if(_errs1 === errors){if(data.firstName !== undefined){const _errs2 = errors;if(typeof data.firstName !== "string"){validate10.errors = [{instancePath:instancePath+"/firstName",schemaPath:"#/properties/firstName/type",keyword:"type",params:{type: "string"},message:"must be string"}];return false;}var valid0 = _errs2 === errors;}else {var valid0 = true;}if(valid0){if(data.lastName !== undefined){const _errs4 = errors;if(typeof data.lastName !== "string"){validate10.errors = [{instancePath:instancePath+"/lastName",schemaPath:"#/properties/lastName/type",keyword:"type",params:{type: "string"},message:"must be string"}];return false;}var valid0 = _errs4 === errors;}else {var valid0 = true;}if(valid0){if(data.age !== undefined){let data2 = data.age;const _errs6 = errors;if(!((typeof data2 == "number") && (isFinite(data2)))){validate10.errors = [{instancePath:instancePath+"/age",schemaPath:"#/properties/age/type",keyword:"type",params:{type: "number"},message:"must be number"}];return false;}var valid0 = _errs6 === errors;}else {var valid0 = true;}}}}}}else {validate10.errors = [{instancePath,schemaPath:"#/type",keyword:"type",params:{type: "object"},message:"must be object"}];return false;}}validate10.errors = vErrors;return errors === 0;};validate10.schema=schema11;",
    "type": "validator",
  },
  {
    "clean": [Function],
    "data": {
      "00014-000": {
        "referenceName": "{{00014-000:symbolName}}",
        "srcPath": "examples/json-schema/example.schema.type.ts",
        "symbolName": "SelfReference",
        "uniqueSymbolName": "{{00014-000:symbolName}}",
      },
      "0005-000": {
        "referenceName": "{{0005-000:symbolName}}",
        "srcPath": "examples/json-schema/example.schema.type.ts",
        "symbolName": "Person",
        "uniqueSymbolName": "{{0005-000:symbolName}}",
      },
      "0009-000": {
        "referenceName": "{{0009-000:symbolName}}",
        "srcPath": "examples/json-schema/example.schema.type.ts",
        "symbolName": "SalesPerson",
        "uniqueSymbolName": "{{0009-000:symbolName}}",
      },
    },
    "prettify": false,
    "relativeSource": "./example.schema",
    "targetPath": "examples/json-schema/schemas/sales-person.schema.js",
    "template": "/**
 * Generated by @skyleague/therefore
 * eslint-disable
 */
"use strict";module.exports = validate10;module.exports.default = validate10;const schema11 = {"$schema":"http://json-schema.org/draft-07/schema#","title":"{{0009-000:uniqueSymbolName}}","allOf":[{"type":"object","properties":{"sales":{"type":"integer"}},"required":["sales"],"additionalProperties":true},{"type":"object","properties":{"firstName":{"type":"string","description":"The person's first name."},"lastName":{"type":"string"},"age":{"type":"number"}},"required":["firstName","lastName","age"],"additionalProperties":true}]};function validate10(data, {instancePath="", parentData, parentDataProperty, rootData=data}={}){let vErrors = null;let errors = 0;const _errs0 = errors;if(errors === _errs0){if(data && typeof data == "object" && !Array.isArray(data)){let missing0;if((data.sales === undefined) && (missing0 = "sales")){validate10.errors = [{instancePath,schemaPath:"#/allOf/0/required",keyword:"required",params:{missingProperty: missing0},message:"must have required property '"+missing0+"'"}];return false;}else {if(data.sales !== undefined){let data0 = data.sales;if(!(((typeof data0 == "number") && (!(data0 % 1) && !isNaN(data0))) && (isFinite(data0)))){validate10.errors = [{instancePath:instancePath+"/sales",schemaPath:"#/allOf/0/properties/sales/type",keyword:"type",params:{type: "integer"},message:"must be integer"}];return false;}}}}else {validate10.errors = [{instancePath,schemaPath:"#/allOf/0/type",keyword:"type",params:{type: "object"},message:"must be object"}];return false;}}var valid0 = _errs0 === errors;if(valid0){const _errs5 = errors;if(errors === _errs5){if(data && typeof data == "object" && !Array.isArray(data)){let missing1;if((((data.firstName === undefined) && (missing1 = "firstName")) || ((data.lastName === undefined) && (missing1 = "lastName"))) || ((data.age === undefined) && (missing1 = "age"))){validate10.errors = [{instancePath,schemaPath:"#/allOf/1/required",keyword:"required",params:{missingProperty: missing1},message:"must have required property '"+missing1+"'"}];return false;}else {if(data.firstName !== undefined){const _errs8 = errors;if(typeof data.firstName !== "string"){validate10.errors = [{instancePath:instancePath+"/firstName",schemaPath:"#/allOf/1/properties/firstName/type",keyword:"type",params:{type: "string"},message:"must be string"}];return false;}var valid2 = _errs8 === errors;}else {var valid2 = true;}if(valid2){if(data.lastName !== undefined){const _errs10 = errors;if(typeof data.lastName !== "string"){validate10.errors = [{instancePath:instancePath+"/lastName",schemaPath:"#/allOf/1/properties/lastName/type",keyword:"type",params:{type: "string"},message:"must be string"}];return false;}var valid2 = _errs10 === errors;}else {var valid2 = true;}if(valid2){if(data.age !== undefined){let data3 = data.age;const _errs12 = errors;if(!((typeof data3 == "number") && (isFinite(data3)))){validate10.errors = [{instancePath:instancePath+"/age",schemaPath:"#/allOf/1/properties/age/type",keyword:"type",params:{type: "number"},message:"must be number"}];return false;}var valid2 = _errs12 === errors;}else {var valid2 = true;}}}}}else {validate10.errors = [{instancePath,schemaPath:"#/allOf/1/type",keyword:"type",params:{type: "object"},message:"must be object"}];return false;}}var valid0 = _errs5 === errors;}validate10.errors = vErrors;return errors === 0;};validate10.schema=schema11;",
    "type": "validator",
  },
]
`;
